pennylane_lightning/core/src/utils/NDPermuter.hpp

```diff
@@ -0,0 +1,268 @@
+ *
+ * The Permuter class represents the front-end interface for calling
+ * permutations, which are a generalization of transposition to high-rank
+ * tensors. The class follows a composition-based approach, where we instantiate
+ * with a given backend permuter, who makes available two `Transpose` methods,
+ * one which returns the transform result, and another which modifies a
+ * reference directly.
+ *
+ *   Example 1:
+ *   const std::vector<size_t> data_in {0,1,2,3,4,5};
+ *   std::vector<size_t> data_out(data_in.size(), 0);
+ *   Permuter<DefaultPermuter<size_t>> p;
+ *   p.Transpose(data_in, {2,3}, data_out, {"a","b"}, {"b","a"});
+ *
+ *   Example 2:
+ *   const std::vector<size_t> data_in {0,1,2,3,4,5};
+ *   Permuter<DefaultPermuter<size_t>> p;
+ *   auto data_out = p.Transpose(data_in, {2,3}, {"a","b"}, {"b","a"});
+ *
+ * @tparam PermuteBackend
+ */
+template <class PermuterBackend> class Permuter {
+  public:
+    /**
+     * @brief Reshape the given lexicographic data vector from old to new index
+     * ordering.
+     *
+     * @tparam T Data participating in the permutation.
+     * @param data_in Input data to be transposed.
+     * @param shape Current shape of the tensor data in each dimension.
+     * @param data_out Output data following the transpose.
+     * @param current_order Current index ordering of the tensor.
+     * @param new_order New index ordering of the tensor.
+     */
+    template <class T>
+    void Transpose(const std::vector<T> &data_in,
+                   const std::vector<size_t> &shape, std::vector<T> &data_out,
+                   const std::vector<std::string> &current_order,
+                   const std::vector<std::string> &new_order) {
+        const std::set<std::string> idx_old(current_order.begin(),
+                                            current_order.end());
+        const std::set<std::string> idx_new(new_order.begin(), new_order.end());
+        const std::size_t data_size = std::accumulate(
+            shape.begin(), shape.end(), 1, std::multiplies<std::size_t>());
+        PL_ABORT_IF_NOT(idx_old.size() == current_order.size(),
+                        "Duplicate existing indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(idx_new.size() == new_order.size(),
+                        "Duplicate transpose indices found. Please ensure "
+                        "indices are unique.");
+        PL_ABORT_IF_NOT(shape.size() == new_order.size(),
+                        "Tensor shape does not match number of indices.");
+        PL_ABORT_IF_NOT(data_size == data_in.size(),
+                        "Tensor shape does not match given input tensor data.");
+        PL_ABORT_IF_NOT(
+            data_size == data_out.size(),
+            "Tensor shape does not match given output tensor data.");
+        PL_ABORT_IF_NOT(
+            idx_old == idx_new,
+            "New indices are an invalid permutation of the existing indices");
```

Comment:
```suggestion
            "New indices are an invalid permutation of the existing indices.");
```

Reply:
To be resolved by later importing directly from JET. Resolving.